zio-http/shared/src/main/scala-2/zio/http/HandlerVersionSpecific.scala

```diff
@@ -0,0 +1,25 @@
+package zio.http
+
+import zio.{Tag, Trace, ZIO, ZLayer}
+
+trait HandlerVersionSpecific {
+  private[http] class ApplyContextAspect[-Env, +Err, -In, +Out, Env0](private val self: Handler[Env, Err, In, Out]) {
+    def apply[Env1: Tag, Ctx: Tag, In1 <: In](aspect: HandlerAspect[Env1, Ctx])(implicit
+      in: Handler.IsRequest[In1],
+      ev: Env0 with Ctx <:< Env,
+      out: Out <:< Response,
+      err: Err <:< Response,
+      trace: Trace,
+      tag: Tag[Env0],
+    ): Handler[Env0 with Env1, Response, Request, Response] =
+      aspect.applyHandlerContext {
+        handler { (ctx: Ctx, req: Request) =>
+          val handler: ZIO[Env, Response, Response] = self.asInstanceOf[Handler[Env, Response, Request, Response]](req)
+          handler.provideSomeLayer[Env0 with Env1](
+            ZLayer.environment[Env0] ++ ZLayer.environment[Env1] ++ ZLayer.succeed(ctx),
+          )
```

Comment:
Might be simplified:

```suggestion
          handler.provideSomeEnvironment[Env0 with Env1](_.add[Ctx](ctx).asInstanceOf[ZEnvironment[Env]])
```

~~`asInstanceOf` was needed in Scala 3. Not sure if Scala 2 can just infer this.~~